### PR TITLE
Add support for basic feature flags

### PR DESCRIPTION
**Pull request by cecinestpasunepipe**
(DIS-1356)

### CODE REVIEW

File: dissect/util/feature.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
+import functools
+import os
+from typing import Callable, Optional
+
+# Register feature flags in a central place to avoid chaos
+F_NOVICE = "novice"
+F_LATEST = "latest"
+F_BETA = "beta"
```
Comment (Schamper) [inline dissect/util/feature.py:8]: Would these not be better as an enum?

File: dissect/util/feature.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
+import functools
+import os
+from typing import Callable, Optional
+
+# Register feature flags in a central place to avoid chaos
+F_NOVICE = "novice"
```
Comment (Schamper) [inline dissect/util/feature.py:6]: I'm not sure I get the purpose of a "novice" feature flag, could you give a usage example?

**Review by Schamper — CHANGES_REQUESTED** [review]

File: dissect/util/feature.py (hunk 2 above)
Comment (cecinestpasunepipe) [reply dissect/util/feature.py]: The novice flag is to shield users from advanced features that might be risky. 

Usage example:

```
@feature(Feature.NOVICE, real_private_keys):
@export(record=PrivateKeyRecord)
    def private_keys(self) -> Iterator[PrivateKeyRecord]:
       print("Are you sure?....")
    .....
```

because user might end up with private keys from client, which might have consequences depending on country/jurisdiction.

Technical issues may also play a role:


```
@feature(Feature.NOVICE, real_central_av_dump):
@export(record=PrivateKeyRecord)
    def private_keys(self) -> Iterator[PrivateKeyRecord]:
        print("This can take a while... and you might acquire too much...")
    .....
```
If a user wants to obtain a central av log and the client of the user uses the same cluster for all databases, the user might end up downloading all DBs of the customer, so the user might need to check this first.

File: dissect/util/feature.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
+import functools
+import os
+from enum import Enum
+from typing import Callable, Optional
+
+
+# Register feature flags in a central place to avoid chaos
+class Feature(Enum):
```
Comment (Schamper) [inline dissect/util/feature.py:8]: If you make it a `Flag` you can OR features together.

File: dissect/util/feature.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    NOVICE = "novice"
+    LATEST = "latest"
+    BETA = "beta"
+
+
+# This set defines the valid flags
+DISSECT_FEATURE_SET = set(item for item in Feature)
+
+# Defines the default flags (as strings)
+DISSECT_FEATURES_DEFAULT = "novice/latest"
```
Comment (Schamper) [inline dissect/util/feature.py:18]: And this could be an OR of the novice and latest flag.

File: dissect/util/feature.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+# Register feature flags in a central place to avoid chaos
+class Feature(Enum):
+    NOVICE = "novice"
+    LATEST = "latest"
+    BETA = "beta"
+
+
+# This set defines the valid flags
+DISSECT_FEATURE_SET = set(item for item in Feature)
```
Comment (Schamper) [inline dissect/util/feature.py:15]: Why is this needed? You can already perform membership checks on Enum/Flag classes.

File: dissect/util/feature.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+
+def feature_disabled_stub() -> None:
+    raise FeatureException("This feature has been disabled.")
+
+
+def feature(flag: Feature, alternative: Optional[Callable] = feature_disabled_stub) -> Callable:
+    """Usage:
+
+    @feature(F_SOME_FLAG, altfunc)
```
Comment (Schamper) [inline dissect/util/feature.py:52]: Outdated docstring.

Can you also update the docstring so that it's formatted nicely in Sphinx?

File: dissect/util/feature.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    Where F_SOME_FLAG is the feature you want to check for and
+    altfunc is the alternative function to serve
+    if the feature flag is NOT set.
+    """
+
+    def decorator(func):
+        if feature_enabled(flag):
+            return func
+        else:
+            return alternative
```
Comment (Schamper) [inline dissect/util/feature.py:64]:
```suggestion
        return func if feature_enabled(flag) else alternative
```

File: dissect/util/feature.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
+import functools
+import os
+from enum import Enum
+from typing import Callable, Optional
+
+
+# Register feature flags in a central place to avoid chaos
+class Feature(Enum):
+    NOVICE = "novice"
```
Comment (Schamper) [inline dissect/util/feature.py:9]: I don't necessarily agree with the feature name "novice", shouldn't it be the other way around? That you guard more advanced features?

**Review by Schamper — CHANGES_REQUESTED** [review]

File: dissect/util/feature.py (hunk 3 above)
Comment (cecinestpasunepipe) [reply dissect/util/feature.py:8]: But then I have to override the __or__/__ror__ myself because the values are strings, or I need to map the strings to the flag first. All of which mean extra code, which might be unnecessary. Are you sure you want this?

File: dissect/util/feature.py (hunk 4 above)
Comment (cecinestpasunepipe) [reply dissect/util/feature.py:18]: See comment above.

File: dissect/util/feature.py (hunk 8 above)
Comment (cecinestpasunepipe) [reply dissect/util/feature.py]: Ok, used "advanced" instead.

File: dissect/util/feature.py (hunk 3 above)
Comment (Schamper) [reply dissect/util/feature.py:8]: Hm you're right. While it feels more correct, it's more effort to achieve the same as you currently do with the split on `/`.

File: dissect/util/feature.py [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+import os
+from enum import Enum
+from typing import Callable, Optional
+
+
+# Register feature flags in a central place to avoid chaos
+class Feature(Enum):
+    ADVANCED = "advanced"
+    LATEST = "latest"
+    BETA = "beta"
```
Comment (Schamper) [inline dissect/util/feature.py:11]:
```suggestion
    LATEST = "latest"
    ADVANCED = "advanced"
    BETA = "beta"
```
Nitpicky, but I believe the "default" flag should be on top.

And maybe experimental over beta? That's mostly personal taste of course, I'm not 100% sure on it either.

File: dissect/util/feature.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    pass
+
+
+@functools.cache
+def feature_flags() -> list[Feature]:
+    return [Feature(name) for name in os.getenv(DISSECT_FEATURES_ENV, DISSECT_FEATURES_DEFAULT).split("/")]
+
+
+@functools.cache
+def feature_enabled(feature: Feature) -> bool:
```
Comment (Schamper) [inline dissect/util/feature.py:31]: Can you add a docstring that includes a code example (if statement) that uses this?

File: dissect/util/feature.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    return feature in feature_flags()
+
+
+def feature_disabled_stub() -> None:
+    raise FeatureException("This feature has been disabled.")
```
Comment (Schamper) [inline dissect/util/feature.py:36]: Could this be more descriptive? Ideally this points to some docs that currently don't exist yet :smile:

File: dissect/util/feature.py [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+
+def feature(flag: Feature, alternative: Optional[Callable] = feature_disabled_stub) -> Callable:
+    """
```
Comment (Schamper) [inline dissect/util/feature.py:40]:
```suggestion
    """Feature flag decorator allowing you to guard a function behind a feature flag.

```
Or something along those lines. This sentence is pretty bad, but there should be a first-line docstring that explains the purpose.

File: dissect/util/feature.py [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+    Usage::
+
+        @feature(Feature.SOME_FLAG, fallback)
+        def my_func( ... ) -> ...
+
+    Where SOME_FLAG is the feature you want to check for and
+    altfunc is the alternative function to serve
```
Comment (Schamper) [inline dissect/util/feature.py:47]:
```suggestion
    Where ``SOME_FLAG`` is the feature you want to check for and ``fallback`` is the alternative function to serve
```

**Review by Schamper — CHANGES_REQUESTED** [review]

**Review by Schamper — APPROVED** [review]
